### PR TITLE
EXPERIMENT Scoped slots for tabs component

### PR DESCRIPTION
**Pull request by DXCanas**
I thought putting components that are only ever meant to be used inside another was weird, so I'm trying a few different ways to go around that approach. This is the one I nagged @christianmemije about on his PR. A little weird, but I wanted to see what you guys thought.

### CODE REVIEW

**Comment by indirectlylit** [comment]
> I thought putting components that are only ever meant to be used inside another was weird

Could you elaborate on this?

I agree that there are pros and cons, but it does at least have precedent in HTML. For example, `<li>` is meant to be used inside `<ul>` or `<ol>` and tables require a `table`, `tr`, `td` nested structure.

File: kolibri/plugins/learn/assets/src/views/index.vue [hunk 1]
```diff
@@ -101,7 +103,6 @@
       'breadcrumbs': require('./breadcrumbs'),
       'search-page': require('./search-page'),
       'tabs': require('kolibri.coreVue.components.tabs'),
-      'tab-link': require('kolibri.coreVue.components.tabLink'),
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/index.vue:104]: I like how this change allows the whole `tabs` component to be used with a single `require`, and the child components simply "come along for the ride".

File: kolibri/plugins/learn/assets/src/views/index.vue [hunk 2]
```diff
@@ -39,26 +39,28 @@
...
-          name="exam-link"
-          v-if="isUserLoggedIn && userHasMemberships"
-          type="icon-and-title"
-          :title="$tr('exams')"
-          icon="assignments"
-          :link="examsLink"
-        />
+        <template scope="tab">
+          <component
+            :is="tab.link"
```
Comment (indirectlylit) [inline kolibri/plugins/learn/assets/src/views/index.vue:44]: I don't like how this change switches to using two advanced vue features - the `:is` attribute on `component` and the `scope` attribute on `template`. I think these might reduce readability more than what's gained by removing the explicit `tab-link` child component reference.

Comment (DXCanas) [reply kolibri/plugins/learn/assets/src/views/index.vue:44]: I agree. I think the HTML argument made above is really good, so we can leave it as is. I'm curious about why we're not just passing in an array or something, though. I understand that this is more explicit, it just seems like we're writing a bunch of repetitious code for no particular reason.[ I liked the way VueJS did it in their example](https://vuejs.org/v2/guide/components.html#Scoped-Slots):
```
<my-awesome-list :items="items">
  <!-- scoped slot can be named too -->
  <template slot="item" scope="props">
    <li class="my-fancy-item">{{ props.text }}</li>
  </template>
</my-awesome-list>
```

Comment (indirectlylit) [reply kolibri/plugins/learn/assets/src/views/index.vue:44]: > I'm curious about why we're not just passing in an array or something, though

I mainly wanted to avoid having complex data structures (e.g. arrays of objects) as props because they're harder to validate and document.

> it just seems like we're writing a bunch of repetitious code for no particular reason

I'm admittedly not the sharpest tool in the shed, but I find scoped slots to be rather confusing. Would they actually reduce code repetition?

Comment (DXCanas) [reply kolibri/plugins/learn/assets/src/views/index.vue:44]: > I mainly wanted to avoid having complex data structures (e.g. arrays of objects) as props because they're harder to validate and document.

That makes sense, it just seems counter intuitive to the idea behind a child component. We have to pass in the same props using the same grandchildren over and over, which makes the template feel more verbose than it needs to be.

 > Would they actually reduce code repetition?

Yes, because you'd only specify the granchild component once for every item in the array. However, you'd have to pass in the array via a prop and the grandchild component via a slot, making it kind of complex.

> I'm admittedly not the sharpest tool in the shed

Nah. It took me a lot of experimenting and rereading to understand how they worked. Christian had some trouble with it too. Maybe this is reason enough to forego the approach.

Basically, adding `scope` to a slot while it's being used in a parent passes up the scope of the child to the parent (only within the slot) which might be anti-pattern.

**Comment by indirectlylit** [comment]
> Basically, adding `scope` to a slot while it's being used in a parent passes up the scope of the child to the parent (only within the slot) which might be anti-pattern.

I don't know if it's an anti-pattern per se, but probably best reserved for cases that really need it since it's harder to reason about. Similarly, I'd argue that even named slots introduce complexity into a component's API, and should be avoided unless totally necessary.
